### PR TITLE
builds: add automatic docker build

### PR DESCRIPTION
**Pull request by diegodelemos**
Closes https://github.com/inveniosoftware/docker-invenio/issues/2 and https://github.com/inveniosoftware/invenio/issues/3804 and https://github.com/inveniosoftware/docker-invenio/issues/4.
Closes https://github.com/inveniosoftware/docker-invenio/issues/7 using first listed option.
Depends on https://github.com/inveniosoftware/docker-invenio/pull/3.

### CODE REVIEW

**Comment by diegodelemos** [comment]
### A possible solution using [Docker Cloud](https://cloud.docker.com/).

1. [Steps](https://hackmd.web.cern.ch/s/HJDgQPKI7) to be followed.
2. See [here](https://github.com/diegodelemos/docker-invenio/pull/1) an example of a pull request on a repository linked with Docker Cloud.
3. Result of merging to master and pushing a tag https://hub.docker.com/r/diegodelemos/invenioland/tags/.

**Advantages**:
- [GitHub integration](https://docs.docker.com/docker-cloud/builds/link-source/#link-to-a-github-user-account).
- Single [configuration file](https://docs.docker.com/docker-cloud/builds/automated-testing/#set-up-automated-test-files).
- Testable

**Disadvantages**:
- Builds are only accessible to logged in people (this is worrying, related to [Docker can not be downloaded if not logged into Docker Store](https://github.com/docker/docker.github.io/issues/6910)).
- The proccess described needs to be replicated for each image (i.e. `inveniosoftware/centos7-python37`, `inveniosoftware/centos7-python36-xrootd`... each of them pointing to the right `Dockerfile`)
- Will it continue to be free? Free option is not even listed in [pricing](https://cloud.docker.com/pricing/), obviously it is implicit but not being clear about is weird.

P.S. I do not have permissions to add the PR to the sprint project.

**Comment by diegodelemos** [comment]
### Using [CentOS container registry](https://registry.centos.org/)

1. Create a [cccp.yaml file in the repository](https://github.com/diegodelemos/docker-invenio/commit/1302e436f80897801e612a276e2ca0a297bf1e62).
2. [Make a PR to CentOS container index](https://github.com/diegodelemos/container-index/commit/febd4c0371e874b988740e5adda2321f906a0bc7) to register the application.

Advantages:
- Scans the image for updates, fixes, capabilities and push it to a public registry (by default, http://registry.centos.org)
- Notifications / alerts regarding scan results (by e-mail) (more info [here](https://github.com/CentOS/container-pipeline-service#how-everything-works)).

Disadvantages:
- More complex/manual setup (see more [here](https://github.com/CentOS/container-index#centos-container-index)). Currently (Thu Aug 23 11:12:00 CEST 2018) the process to have this working would mean:
    - [Make a PR to CentOS container index](https://github.com/diegodelemos/container-index/commit/febd4c0371e874b988740e5adda2321f906a0bc7) to add Invenio.
    - On every release, make another PR, see for example how is done for [mattermost](https://github.com/CentOS/container-index/blob/master/index.d/mattermost.yaml).
    - Follow this [steps](https://wiki.centos.org/QaWiki/CI/GettingStarted) in order to setup Centos CI based on Jenkins.
     <details>
  <summary>Quick summary about the steps</summary>

  * Create a bug request on https://bugs.centos.org , in the "Buildsys" Project, and the "Ci.centos.org Ecosystem Testing" category

  * Add some description and details about the project (and SIG name if a SIG)
  * Attach a ssh public key:
    * That key will be used to let you enter the CI environment (and for example the node where your Jenkins slave will be hosted)
    * That key will also be added to every node provisioned/given to you through the Duffy middleware (see http://wiki.centos.org/QaWiki/CI/Duffy)

  * (optional) Attach a gpg public key (or link to it) : that can permit sending some informations (like credentials) through gpg encrypted mail, if needed
  * Wait for a SIG member/leader (if for a SIG) to approve or if a different project, wait for someone from the CI core team to validate the request.
  * Make sure you sign up for the ci-users list ( https://lists.centos.org/mailman/listinfo/ci-users )
</details>

**Comment by diegodelemos** [comment]
@tiborsimko looks like travis is a requirement for the build but there is no `.travis.yml` in the repo, can you disable it since the image is already tested in the Docker Cloud build?

**Comment by diegodelemos** [comment]
### Final solution

Since [Docker Cloud is it going to be free only during its beta face](https://docs.docker.com/docker-cloud/builds/automated-build/#configure-automated-build-settings) (see point 5), we do, as the simplest solution:

- [x] Switch the Docker Cloud CI step to be a simple [Travis Docker build](https://docs.travis-ci.com/user/docker/#building-a-docker-image-from-a-dockerfile) (done see [commit](https://github.com/inveniosoftware/docker-invenio/pull/6/commits/f5bf49eb1185b640d2ba5c3f7b94063b4c1eed24) and [build passing](https://travis-ci.org/diegodelemos/docker-invenio/builds/419648234)).
- [x] Use [Docker Hub builds](https://docs.docker.com/docker-hub/builds/) to keep images up to date when we do any change that is merged into master (done see [here](https://hub.docker.com/r/inveniosoftware/centos7-python/builds/))
- [x] Use [Docker Hub repository links](https://docs.docker.com/docker-hub/builds/#repository-links) to keep up to date with `centos:7` base image (done, only visible with admin rights).

P.S. https://travis-ci.org/inveniosoftware/docker-invenio needs to be activated.

File: .travis.yml
Comment (tiborsimko) [inline .travis.yml]: Perhaps you can use the full image name here:

```
docker build -t inveniosoftware/centos7-python ...
```

for the sake of consistency, even though the image won't be pushed by Travis.


BTW speaking of https://hub.docker.com/u/inveniosoftware/ I don't think `centos7-python` is a good base image name, because:

- based on the name alone, seeing `centos7-python:3.7`, people may think we are providing a rather general customised CentOS7 base image adding Python-3.7 only -- while the image does several more Invenio-specific things, such as installing `npm`, which people may not want;

- if we need to differentiate one day between `npm6` and `npwm8`, then how would be expressed in the image? via different name? via different tag?

- if we make mistake and need to repush several different versions of the Invenio 3.1.2 base image using Python 3.8, how would this be represented in the tag?  3.8.1-N where N is our local "build" version, so to speak?

- imagine that Invenio 3.2 or other such later version will change something in the way how npm is used, or how the instance directories are created; in this case, both Invenio 3.0 and Invenio 3.2 release series will be runnable on Python 3.7. How would the two base images be distinguishable?

I think a more future-proof naming schema would be something like:

```
invenio-centos7-base:3.0.0
invenio-centos7-base:3.0.0-py37
invenio-centos7-base:3.2.0-py37-npm6
invenio-centos7-base:3.2-py37-npm8
```

or other such alternative naming scheme such as `invenio3-centos7-py36:1.0.0` which would however lead to a proliferation of various image names instead of just tags.


This should better express that (a) the image is not general Python customisation but meant as a base image for Invenio instances, (b) that it is suitable for some particular Invenio release series version, and (c) that it customises more than py37 should a need arise one day to distinguish between npm6 and npm8, say.

**Review by tiborsimko — APPROVED** [review]

File: .travis.yml
Comment (tiborsimko) [reply .travis.yml]: For the sake of further discussion about how to choose tags, it may be interesting to recall that we need to cover various upgrade scenarios: (1) Python version upgrades e.g. 3.7.0->3.7.1; (2) CentOS7 based image upgrades e.g. security; (3) Invenio dependency updates e.g. Invenio 3.0.x necessitating different directory structure or different npm package than Invenio 3.1.x; (4) incremental build number (or SHA1) for the image to cover for our own packaging/building errors.  

Basically, we are packaging Python-on-CentOS for the use of various Invenio installations that may be on various versions etc, so we are doing the work here that is very similar to what a typical GNU/Linux distribution packager does (e.g. `linux-headers 4.18.5.arch1-1`, `2.6.0-1ubuntu1`, etc).

Having well-though and extensible tag structure is important for any production service that may be based on these images and that deploys latest upgrades and that need to quickly roll back in case of unforeseen problems.

Examples to illustrate discussion:

```
FROM inveniosoftware/centos7-python:3.6
FROM inveniosoftware/centos7-python:3.7

FROM inveniosoftware/centos7-python:3.6-invenio30
FROM inveniosoftware/centos7-python:3.6-invenio30
FROM inveniosoftware/centos7-python:3.7-invenio32
FROM inveniosoftware/centos7-python:3.7-invenio32

FROM inveniosoftware/centos7-python:3.7.0-invenio30
FROM inveniosoftware/centos7-python:3.7.1-invenio30

FROM inveniosoftware/centos7-python:3.7.2-invenio30-1
FROM inveniosoftware/centos7-python:3.7.2-invenio30-2

FROM inveniosoftware/centos7-python:3.7.0-invenio30-3025579
FROM inveniosoftware/centos7-python:3.7.1-invenio30-26ccb52
```

**Pull request merged**
